### DISCUSSION TITLE
Restoring in memory state when outage of nodes occurs

### DISCUSSION THREAD

**Discussion by Patrick354** [ava-labs/hypersdk#925] [Chat]
In the tokenvm example, in the case of an outage on all of the running nodes, how can we rebuild the orderbook that is in memory since the nodes are non-archival ?

**Comment by patrick-ogrady** [comment]
I think the big decision here is whether to push most "user interaction" to an integrated RPC (like most blockchains today) or to require anyone running a HyperVM in production to maintain some external data pipeline/API.

In the first case, a HyperVM would iterate over (potentially all) state to re-populate any in-memory data structures required to service queries. In the latter case, state diffs/results streaming to a data pipeline should provide a robust view (assuming the HyperSDK ensures that every block is streamed at least once).

Based on my experience running nodes, I'm partial to the external data pipeline for anything that isn't querying state directly. But supporting state iteration shouldn't be that bad for anyone that just wants to scale their API by running more nodes (we've just found that to be very expensive).
